sos/plugins/juju_controller.py

```diff
@@ -0,0 +1,157 @@
+            collections = collections.replace(search, replace)
+        collections = collections.split()
+
+        # We don't collect txns.log, statuseshistory, metrics, txns due to size
+        collections_ignore = ['txns.log', 'statuseshistory', 'metrics', 'txns']
+        for item in collections_ignore:
+            if item in collections:
+                collections.remove(item)
+
+        for collection in collections:
+            filename = "mongoexport_collection_{}.json".format(collection)
+            # This will execute a mongoexport command, note it will add the
+            # password to the sos log, however we remove it from the filename
+            mongoexport_cmd = (
+                "mongoexport --host 127.0.0.1 --port 37017 --db juju "
+                + "--authenticationDatabase admin --ssl "
+                + "--sslAllowInvalidCertificates --username '{}' --password "
+                + "'{}' --collection {} --jsonArray"
+            )
+            self.add_cmd_output(
+                mongoexport_cmd.format(juju_machine_id, dbpass, collection),
+                suggest_filename=filename,
+            )
+
+    def get_juju_machine_id(self):
+        # get the juju machine name machine-X
+        for dirpath, dirnames, filenames in os.walk("/var/lib/juju/agents"):
+            for dirname in dirnames:
+                match = re.match(r"(machine-\d+)", dirname)
+                if match:
+                    return match.group(1)
+        return None
+
+    def get_jujudb_pass(self, juju_machine_id):
+        # Get agent config and figure out password
+        try:
+            fp = open(
+                "/var/lib/juju/agents/" + juju_machine_id + "/agent.conf", "r"
+            )
+            filecontents = fp.read()
+            fp.close()
+        except Exception:
+            return None
+        for line in iter(filecontents.splitlines()):
+            match = re.match(r"^statepassword: (\S+)", line)
+            if match:
+                return match.group(1)
+        return None
+
+    def add_query(self, juju_machine_id, dbpass, queryname, query):
+        # Run a command for a single query
+        filename = "mongo_{}.json".format(queryname)
+        # This will execute a mongo command, note it will add the password to
+        # the sos log we remove it from the filename however
+        mongocmd = (
+            "mongo 127.0.0.1:37017/juju --authenticationDatabase admin"
+            + " --ssl --sslAllowInvalidCertificates --quiet --username '{}' "
+            + "--password '{}' --eval '{}'"
+        )
+        self.add_cmd_output(
+            mongocmd.format(juju_machine_id, dbpass, query),
+            suggest_filename=filename,
+        )
+
+    def setup(self):
+
+        self.add_copy_spec(
+            [
+                "/etc/default/mongodb",
+                "/var/lib/juju/bootstrap-params",
+                "/lib/systemd/system/juju-db/juju-db.service",
+            ]
+        )
```

Comment:
Stylistic nitpick: Keep the brackets on the same line.

```python
self.add_copy_spec([
    'foo',
    'bar'
])
```

Reply:
Fixed on 9c909bd